src/gdb_server/src/util.rs

```diff
@@ -0,0 +1,330 @@
+
+const CR0_PG: u64 = 0x80000000;
+const CR4_PAE: u64 = 0x00000020;
+const CR4_LA57: u64 = 0x00001000;
+const EFER_LME: u64 = 0x00000100;
+const PDPTE_PS: u64 = 0x00000080;
+const PDE_PS: u64 = 0x00000080;
+
+const BIT_P: u64 = 0x1;
+
+// [PML4E, PDPTE_PS0, PDPTE_PS1, PDE_PS0, PDE_PS1, PTE]
+const TABLE_ENTRY_RSVD_BITS: [u64; 6] = [0x80, 0x0, 0x3fffe000, 0x0, 0x1fe000, 0x0];
+
+const TABLE_ENTRY_MASK: u64 = 0x000ffffffffff000u64;
+
+#[derive(PartialEq, Eq, Debug)]
+enum PagingType {
+    NONE,
+    _32BIT,
+    PAE,
+    _4LVL,
+    _5LVL,
+}
+
+pub enum DebugEvent {
+    START,
+    NOTIFY(FullVcpuState),
+    GET_REGS,
+    PEEK_REGS(FullVcpuState),
+    CONTINUE(bool),
+    STEP_INTO(bool),
+}
+
+#[derive(Default)]
+pub struct FullVcpuState {
+    pub regular_regs: kvm_regs,
+    pub special_regs: kvm_sregs,
+}
+
+#[derive(Debug)]
+pub enum DebuggerError {
+    InvalidState,
+    ChannelError,
+    MemoryError,
+    IoctlError(Error),
+    InvalidLinearAddress,
+    UnsupportedPagingStrategy,
+}
+
+impl Display for DebuggerError {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        match self {
+            DebuggerError::InvalidState => write!(f, "[GDB Server] An invalid state was reached"),
+            DebuggerError::ChannelError => write!(
+                f,
+                "[GDB Server] An error interrupted the GDB server - main thread communication"
+            ),
+            DebuggerError::MemoryError => write!(f, "[GDB Server] Failed access to guest memory"),
+            DebuggerError::IoctlError(errno) => {
+                write!(f, "[GDB Server] Failed ioctl call: {}", errno)
+            }
+            DebuggerError::InvalidLinearAddress => write!(
+                f,
+                "[GDB Server] An invalid linear address was passed from client"
+            ),
+            DebuggerError::UnsupportedPagingStrategy => write!(
+                f,
+                "A paging strategy that is not currently supported has been detected"
+            ),
+        }
+    }
+}
+
+pub struct Debugger;
+
+impl Debugger {
+    pub fn enable_kvm_debug(vcpu: &VcpuFd, step: bool) -> Result<(), DebuggerError> {
+        let mut control: __u32 = 0;
+        control |= KVM_GUESTDBG_ENABLE | KVM_GUESTDBG_USE_SW_BP;
+        if step {
+            control |= KVM_GUESTDBG_SINGLESTEP;
+        }
+        let debug_struct = kvm_guest_debug {
+            control: control,
+            pad: 0,
+            arch: kvm_guest_debug_arch {
+                debugreg: [0, 0, 0, 0, 0, 0, 0, 0],
+            },
+        };
+
+        if let Err(errno) = vcpu.set_guest_debug(&debug_struct) {
+            return Err(DebuggerError::IoctlError(errno));
+        }
+
+        Ok(())
+    }
+
+    /// Function that performs guest page-walking, on top of the 4-Level paging mechanism,
+    /// obtaining the physical address corresponding to the one received from a GDB client
+    pub fn virt_to_phys(
```

Comment:
This function is quite long and the algorithm overall should be explained in a comment here.

We could also break the function down in smaller pieces so that it's easier to understand it from the conceptual point of view.